### DISCUSSION TITLE
Setting state for array of nested objects

### DISCUSSION THREAD

**Discussion by romankoho** [pmndrs/zustand#1698] [Q&A]
I'm new to react (and zustand) and can't get this working

so I have a (simplified) class "Room":

```
class Room {
    id: string,
    name: string,
    tags: string []
}  
```

then I created a store:
```

type RoomsStore = {
    rooms: Room[];
    setRooms: (updatedRooms: Rooms[]) => void
}

const useRoomsStore = create<RoomsStore>()(
   (set) => ({
        rooms: [],
        setRooms: (newRooms: Room[]) => set((state) => ({ rooms: newRooms })),
}));
```

with setRooms it seems like I don't update the state correctly. The id and name of all objects in the array update properly but the tags[] array within the single objects in the rooms array do not update. I looked into immer as well but I still struggle (also with the syntax).

Would be great if someone can help me.
Cheers

**Comment by dai-shi** [comment]
I suspect you are mutating `tags`. You should replace the array.
```js
// doesn't work
tags.push('foo')
// does work
tags = [...tags, 'foo']
```

(And, then you should replace the entire room object, actually. If you feel this isn't for you, check out [valtio](https://github.com/pmndrs/valtio).)
